src/main/java/org/mule/modules/cors/ValidateMessageProcessor.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright 2014
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mule.modules.cors;
+
+import org.mule.DefaultMuleEvent;
+import org.mule.NonBlockingVoidMuleEvent;
+import org.mule.OptimizedRequestContext;
+import org.mule.api.MessagingException;
+import org.mule.api.MuleEvent;
+import org.mule.api.MuleException;
+import org.mule.api.MuleMessage;
+import org.mule.api.store.ObjectStoreException;
+import org.mule.api.transport.ReplyToHandler;
+import org.mule.processor.AbstractRequestResponseMessageProcessor;
+
+import org.apache.commons.lang.StringUtils;
+
+public class ValidateMessageProcessor extends AbstractRequestResponseMessageProcessor {
+
+    private boolean publicResource;
+    private boolean acceptsCredentials;
+    private CorsConfig config;
+
+    @Override
+    protected MuleEvent processRequest(MuleEvent event) throws MuleException {
+        if (publicResource && acceptsCredentials) {
+            throw new IllegalArgumentException("Resource may not be public and accept credentials at the same time");
+        }
+
+        //read the origin
+        String origin = event.getMessage().getInboundProperty(Constants.ORIGIN);
+
+        //if origin is not present, then not a CORS request
+        if (StringUtils.isEmpty(origin)) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Request is not a CORS request.");
+            }
+            return event;
+        }
+
+        //read headers including those of the preflight
+        String method = event.getMessage().getInboundProperty(Constants.HTTP_METHOD);
+
+        //decide if we want to invoke the flow.
+        if (shouldInvokeFlow(origin, method, publicResource)) {
+            return event;
+        }
+
+        //setting the response to null.
+        event.getMessage().setPayload(null);
+        //Set next message processor to null to stop further processing
+        setListener(null);
```

Comment:
You shouldn't modify the listener as it will affect concurrent and subsequent requests. 
Returning null avoids executing next chain or you could add some conditional logic before calling processNext.